clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java

```diff
@@ -0,0 +1,142 @@
+        assertFalse("Expected connection not to be blocked during an ongoing connection attempt", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertFalse("Expected connection to have no ready nodes.", connectionStates.hasReadyNodes());
+
+        time.sleep(100);
+
+        // Successful connection
+        connectionStates.ready(nodeId);
+        assertEquals("Expected connection state to be READY after successful connection.", connectionStates.connectionState(nodeId), ConnectionState.READY);
+        assertTrue("Expected connection to be ready after successful connection.", connectionStates.isReady(nodeId));
+        assertTrue("Expected connection to have nodes ready to send requests to after successful connection. ", connectionStates.hasReadyNodes());
+        assertFalse("Expected connection not to remain in connecting state after successful connection.", connectionStates.isConnecting(nodeId));
+        assertFalse("Expected block on connection to be removed after successful connection.", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertEquals("Expected backoff value on established connection to be Long.MAX_VALUE", connectionStates.connectionDelay(nodeId, time.milliseconds()), Long.MAX_VALUE);
+
+        time.sleep(15000);
+
+        // Disconnected from broker
+        connectionStates.disconnected(nodeId, time.milliseconds());
+        assertEquals("Expected connection state to be DISCONNECTED after disconnect event.", connectionStates.connectionState(nodeId), ConnectionState.DISCONNECTED);
+        assertTrue("Expected connection to be disconnected after disconnect event.", connectionStates.isDisconnected(nodeId));
+        assertTrue("Expected connection to be blocked after disconnect event.", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertFalse("Expected connection not to still be in connecting state after disconnect event.", connectionStates.isConnecting(nodeId));
+        assertFalse("Expected connection to have no ready nodes after disconnect event.", connectionStates.hasReadyNodes());
+        assertFalse("Expected connection to be unable to connect immediately after disconnect event.", connectionStates.canConnect(nodeId, time.milliseconds()));
+
+        // After disconnecting we expect a backoff value equal to the reconnect.backoff.ms setting (plus minus 20% jitter)
+        long lowerLimit = Math.round(reconnectBackoffMsTest * (1 - reconnectBackoffJitter));
+        long upperLimit = Math.round(reconnectBackoffMsTest * (1 + reconnectBackoffJitter));
+        long currentBackoff = connectionStates.connectionDelay(nodeId, time.milliseconds());
+        assertTrue("Expected initial backoff after disconnect to be within 20% jitter of 'reconnect.backoff.ms' parameter.", lowerLimit <= currentBackoff && currentBackoff <= upperLimit);
```

Comment:
If you use the `assertEquals` that takes a `double`, you can pass a `delta` value, which makes the code a lot more concise.